src/policy/satisfy.rs

```diff
@@ -38,208 +39,135 @@ impl<'a, Pk: ToPublicKey> Satisfier<Pk> for PolicySatisfier<'a, Pk> {
-                    }
-
-                    SatisfierExtData {
-                        witness: VecDeque::new(),
-                        witness_cost: 0,
-                        program: sub_policies[i].compile(),
-                        program_cost: 0,
+                // Sort by witness cost and mark everything except the cheapest k as to-prune
+                let mut sorted_costs: Vec<_> = costs.iter().copied().enumerate().collect();
+                sorted_costs.sort_by_key(|pair| pair.1);
+                let b1 = Arc::new(Value::u1(1));
+                let mut threshold_failed = false;
+                for &(idx, _) in &sorted_costs[..k] {
+                    if nodes[idx].must_prune() {
+                        // Unlike in the `or` case, where two pruned branches will automatically
+                        // cause the `or` itself to be pruned, with thresholds we have to track
+                        // this manually.
+                        threshold_failed = true;
                     }
+                    witness_bits[idx] = Some(Arc::clone(&b1));
                 }
-
-                /// Return summand program for `i`th child
-                fn get_summand<Pk: ToPublicKey>(
-                    i: usize,
-                    satisfiable_children: &mut Vec<(SatisfierExtData, usize)>,
-                    sub_policies: &[Policy<Pk>],
-                ) -> SatisfierExtData {
-                    let mut sat = get_sat_or_default(i, satisfiable_children, sub_policies);
-                    // Satisfactions always have non-zero program cost
-                    let branch = if sat.program_cost == 0 {
-                        sat.witness.push_front(Value::u1(0));
-                        UsedCaseBranch::Right
-                    } else {
-                        sat.witness.push_front(Value::u1(1));
-                        UsedCaseBranch::Left
-                    };
-                    sat.witness_cost += 1;
-                    sat.program = compiler::thresh_summand(sat.program, branch).unwrap();
-                    sat.program_cost += 9;
-
-                    sat
+                for &(idx, _) in &sorted_costs[k..] {
+                    nodes[idx] = nodes[idx].pruned();
                 }
 
-                let mut sat = get_summand(0, &mut satisfiable_children, sub_policies);
-
-                for i in 1..sub_policies.len() {
-                    let child_sat = get_summand(i, &mut satisfiable_children, sub_policies);
-
-                    sat.witness.extend(child_sat.witness);
-                    sat.witness_cost += child_sat.witness_cost;
-                    sat.program = compiler::thresh_add(sat.program, child_sat.program).unwrap();
-                    sat.program_cost += child_sat.program_cost + 6;
+                let k = u32::try_from(k).expect("k less than 2^32");
+                if threshold_failed {
+                    super::serialize::threshold(k, &nodes, &witness_bits).pruned()
```

Comment:
3030b553f76c1a51e79bd21cba68ed8e987b6565: If we already know that the threshold fails, why construct it in the first place? Could we not do `super::serialize::trivial().pruned()`?

This doesn't work for the other occurrences of `.pruned()` because these expressions come from the sub policies and are already constructed.

Reply:
No, `trivial` has a completely different CMR than the actual threshold. We can't replace branches with unrelated ones like that.

Reply:
Ah, makes sense. I guess constructing the threshold is not too bad if we have already constructed the children.

Reply:
Maybe let's file an issue and see if we can make a `WitnessNode` constructor that produces a `must_prune` node with a specified CMR, and an inner of `Unit` ... I need to think a bit but I think that'd be fine, and would let us do this.

Or don't file an issue, and I'll try to remember to do this :). But because it's an optimization I'd like to do it after this PR, which is otherwise a bit big.

Reply:
We need to know the CMR first. And we have to construct the expression for that, don't we?

Reply:
Ah, yes, you're right.

Though we could cache it by changing `Policy` to be a wrapper type around a `Fragment` enum ;)

Reply:
Actually, that would be quite hard/annoying to do, because it would require `ToPublicKey` to compute the cached CMR, and we don't assume `ToPublicKey` until we're actually trying to do a satisfaction..